experimental/zerotrie/src/cursor.rs

```diff
@@ -0,0 +1,236 @@
+    ///
+    /// Find the longest prefix match:
+    ///
+    /// ```
+    /// use zerotrie::ZeroTrieSimpleAscii;
+    ///
+    /// // A trie with two values: "abc" and "abcdef"
+    /// let trie = ZeroTrieSimpleAscii::from_bytes(b"abc\x80def\x81");
+    ///
+    /// // Find the longest prefix of the string "abcdxy":
+    /// let query = b"abcdxy";
+    /// let mut longest_prefix = 0;
+    /// let mut cursor = trie.cursor();
+    /// for (i, b) in query.iter().enumerate() {
+    ///     // Checking is_empty() is not required, but it is
+    ///     // good for efficiency
+    ///     if cursor.is_empty() {
+    ///         break;
+    ///     }
+    ///     if cursor.value().is_some() {
+    ///         longest_prefix = i;
+    ///     }
+    ///     cursor.step(*b);
+    /// }
+    ///
+    /// // The longest prefix is "abc" which is length 3:
+    /// assert_eq!(longest_prefix, 3);
+    /// ```
+    #[inline]
+    pub fn cursor(&self) -> ZeroTrieSimpleAsciiCursor {
+        ZeroTrieSimpleAsciiCursor {
+            trie: self.as_borrowed_slice(),
+        }
+    }
+}
+
+impl<'a> ZeroTrieSimpleAscii<&'a [u8]> {
+    /// Same as [`ZeroTrieSimpleAscii::cursor()`] but moves self to avoid
+    /// having to doubly anchor the trie to the stack.
+    #[inline]
+    pub fn into_cursor(self) -> ZeroTrieSimpleAsciiCursor<'a> {
+        ZeroTrieSimpleAsciiCursor { trie: self }
+    }
+}
+
+/// A cursor into a [`ZeroTrieSimpleAscii`], useful for stepwise lookup.
+///
+/// For examples, see [`ZeroTrieSimpleAscii::cursor()`].
+#[derive(Debug)]
+pub struct ZeroTrieSimpleAsciiCursor<'a> {
+    trie: ZeroTrieSimpleAscii<&'a [u8]>,
+}
+
+impl<'a> ZeroTrieSimpleAsciiCursor<'a> {
+    /// Steps the cursor one byte into the trie.
+    ///
+    /// # Examples
+    ///
+    /// Unrolled loop checking for string presence at every step:
+    ///
+    /// ```
+    /// use zerotrie::ZeroTrieSimpleAscii;
+    ///
+    /// // A trie with two values: "abc" and "abcdef"
+    /// let trie = ZeroTrieSimpleAscii::from_bytes(b"abc\x80def\x81");
+    ///
+    /// // Search the trie for the string "abcdxy"
+    /// let mut cursor = trie.cursor();
+    /// assert_eq!(cursor.value(), None); // ""
+    /// cursor.step(b'a');
+    /// assert_eq!(cursor.value(), None); // "a"
+    /// cursor.step(b'b');
+    /// assert_eq!(cursor.value(), None); // "ab"
+    /// cursor.step(b'c');
+    /// assert_eq!(cursor.value(), Some(0)); // "abc"
+    /// cursor.step(b'd');
+    /// assert_eq!(cursor.value(), None); // "abcd"
+    /// assert!(!cursor.is_empty());
+    /// cursor.step(b'x'); // no strings have the prefix "abcdx"
+    /// assert!(cursor.is_empty());
+    /// assert_eq!(cursor.value(), None); // "abcdx"
+    /// cursor.step(b'y');
+    /// assert_eq!(cursor.value(), None); // "abcdxy"
+    /// ```
+    #[inline]
+    pub fn step(&mut self, byte: u8) {
+        step_bsearch_only(&mut self.trie.store, byte)
+    }
+
+    /// Takes the value at the current position and moves the cursor.
```

Comment:
it's unclear to me *where* it moves the cursor, I don't see a default place where it could be moved

Reply:
I deleted `peek_value` which means I no longer need that clause to distinguish the behavior of `value` from `take_value`.

Reply:
Ok, but why does this "take" and use a `&mut`? This should not mutate.

Reply:
Values are stored in trie nodes. When we read a value, we can step over that trie node so that we don't need to do that again next time we call `step` with the next character.

Reply:
Is `value()` idempotent?

Reply:
As the docs example demonstrates, the first time you call `.value()`, you get the value out, and from that point forward it returns `None`

Reply:
Can the cursor cache the value?

Reply:
I consider reading from a ZeroTrie to be performance-critical code so I favor a well documented solution that optimizes for performance, in this case not duplicating the work of reading the value node multiple times.

Caching the value adds extra branches which is not necessary. If you don't want to lose the value, you can either cache it externally or call `.clone().value()` which is equivalent to the `.peek_value()` function I deleted.

Reply:
Here's another mental model for why this mutates: it is removing the empty string from the trie, returning the value associated with the empty string. Then, the next time you call the function, the empty string is no longer in the trie, so you get back `None`.

Reply:
then this should be called `take` or `remove`

Reply:
I can change the function name to `take_value()`.

Reply:
That doesn't work well in the prefix match example which wants to check the presence of a value after each step.

Reply:
`pub fn value(self) -> Result<usize, Self>`?

Reply:
I could make a moving `pub fn value(self)` along with a borrowing `pub fn has_value(&self)` which is _mostly_ as efficient as `pub fn take_value(&mut self)` but I still prefer the mutating value function.

Reply:
> `pub fn value(self) -> Result<usize, Self>`?

That signature doesn't allow checking the presence of a value and then continuing, as we need to do in the longest-prefix example.

Reply:
I changed the function name to `take_value`